torch/csrc/jit/passes/memory_planning/linear_scan.cpp

```diff
@@ -0,0 +1,147 @@
+    // that alloc to the end of the current allocs)
+    // 3. brand new alloc all the way at the end of the currently allocated
+    // memory space
+
+    auto candidate_reg = std::find_if(
+        avail_regions.begin(),
+        avail_regions.end(),
+        [&aligned_curr_size](auto avail_reg) {
+          return avail_reg.size >= aligned_curr_size;
+        });
+
+    if (candidate_reg != avail_regions.end()) {
+      size_t candidate_offset = candidate_reg->offset;
+      size_t candidate_size = candidate_reg->size;
+
+      // note this erases at the iter position (i.e. not all matching regions,
+      // of which there might be multiple)
+      avail_regions.erase(candidate_reg);
+      active[curr_range] = {candidate_offset, aligned_curr_size};
+      // if candidate_reg is bigger than we need then we can
+      // split it and keep that available space for other live ranges
+      if (candidate_size - aligned_curr_size > 0) {
+        avail_regions.insert(MemRegion{
+            candidate_offset + aligned_curr_size,
+            candidate_size - aligned_curr_size});
+      }
+      continue;
+    }
+
+    TORCH_INTERNAL_ASSERT(
+        candidate_reg == avail_regions.end(), "inconsistent state");
+
+    // no avail regions - look for lvr swap candidate in reverse order of
+    // endpoint (i.e. latest ending)
+    auto swap_candidate = std::find_if(
+        active.rbegin(),
+        active.rend(),
+        [&curr_range, &aligned_curr_size](auto item) {
+          return item.first.lvr.end > curr_range.lvr.end &&
+              item.second.size >= aligned_curr_size;
+        });
+
+    if (swap_candidate != active.rend()) {
+      auto candidate_reg = swap_candidate->second;
+      size_t candidate_offset = candidate_reg.offset;
+      size_t candidate_size = candidate_reg.size;
+      active[curr_range] = {candidate_offset, aligned_curr_size};
+      // split region (potentially)
+      if (candidate_size > aligned_curr_size) {
+        avail_regions.insert(MemRegion{
+            candidate_offset + aligned_curr_size,
+            candidate_size - aligned_curr_size});
+      }
+
+      // insert spilled lvr/region at the end
+      active[swap_candidate->first] = {curr_end_offset, candidate_size};
+      curr_end_offset += candidate_size;
+      continue;
+    }
+
+    TORCH_INTERNAL_ASSERT(
+        candidate_reg == avail_regions.end() && swap_candidate == active.rend(),
+        "inconsistent state");
+
+    // create new alloc
+    active[curr_range] = {curr_end_offset, aligned_curr_size};
+    curr_end_offset += aligned_curr_size;
+  }
+
+  // expire any remaining intervals
+  allocate_inactive_ranges(
+      {std::numeric_limits<size_t>::max(), std::numeric_limits<size_t>::max()});
+
+  std::vector<MemAllocation> allocations;
+  allocations.reserve(allocated_ranges.size());
+  for (const auto& item : allocated_ranges) {
+    allocations.push_back({item.first, item.second});
```

Comment:
err this looks fishy... you already reserved enough elements for `allocated_ranges` but then you are pushing back those on top reserved ones?

Reply:
discussed offline